src/libraries/System.Runtime/tests/System/GCTests.cs

```diff
@@ -1058,11 +1058,52 @@ private static void AllocateArrayTooLarge()
             Assert.Throws<OutOfMemoryException>(() => GC.AllocateUninitializedArray<double>(int.MaxValue, pinned: true));
         }
 
-        [Fact]
-        private static void AllocateArrayRefType()
+        struct EmbeddedValueType<T>
+        {
+            unsafe fixed byte _[7];
```

Comment:
```suggestion
            // Adding a simple field to force non-zero object offsets
            int _;
```

I assume the above is sufficient for these purposes.

Reply:
A single byte might fulfil the same purposes. Not sure about whether the VM will actually rearrange references fields to the start, actually. Or is that only for auto-layout (not default for structs)?

Reply:
> Or is that only for auto-layout (not default for structs)?

Managed structs always have Auto layout (or Explicit).

Reply:
> > Or is that only for auto-layout (not default for structs)?
> 
> Managed structs always have Auto layout (or Explicit).

That is not true. C# `struct`s or .NET ValueTypes, by default, are marked as `LayoutKind.Sequential`. C# `class`s or .NET types derived from `System.Object` are marked as `LayoutKind.Auto` by default.

Reply:
> are marked as `LayoutKind.Sequential`

Yes but the VM ignores this, see this sample:
```C#
[StructLayout(LayoutKind.Sequential)] struct S1 { public nint a; public object b; }
[StructLayout(LayoutKind.Sequential)] struct S2 { public nint a; public nint b; }
S1 s = new S1 { a = -1, b = null };
Console.WriteLine(Unsafe.As<S1, S2>(ref s).b);
```
It prints -1, which indicates that in S1 field b was placed first.
If you do:
```C#
[StructLayout(LayoutKind.Explicit)] struct S1 { [FieldOffset(0)] public nint a; [FieldOffset(8)] public object b; }
[StructLayout(LayoutKind.Explicit)] struct S2 { [FieldOffset(0)] public nint a; [FieldOffset(8)] public nint b; }
S1 s = new S1 { a = -1, b = null };
Console.WriteLine(Unsafe.As<S1, S2>(ref s).b);
```
You'll then get the correct 0 which indicates that Explicit layout is respected.

Reply:
> Yes but the VM ignores this, see this sample:

The VM absolutely does not ignore this, it is incredibly important for type layouts and a core concept in interop. The example above is fundamentally flawed because the type is non-blittable. Non-blittable types are not normally involved in interop scenarios therefore things can change. This PR is making an assumption on that, which has been deemed fine and calculated. However, making that statement based on a single observation is flawed reasoning.

See https://github.com/dotnet/runtime/blob/531ad9596fef94194744ccfcac7159d1c1ae5e57/src/coreclr/vm/methodtablebuilder.cpp#L1760 and https://github.com/dotnet/runtime/blob/531ad9596fef94194744ccfcac7159d1c1ae5e57/src/coreclr/vm/methodtablebuilder.cpp#L11565.

Reply:
> This PR is making an assumption on that, which has been deemed fine and calculated.

We actually might want to validate the layout scenarios however. If the type loader is altering layouts this PR becomes a bit more controversial.

Reply:
I should also apologize for being a overly aggressive on this one. I'm sorry. The statement out of context is what I was responding to and that is wrong.

If non-blittable valuetypes don't respect the sequential layout, then I don't fully understand how this sort of offset indexing will work reliably. It would appear we would be at the mercy of the typeloaders layout algorithm, which has been known to change. This actually might be worth considering a bug that needs to be fixed. I defer to the type system people though.

Reply:
> The example above is fundamentally flawed because the type is non-blittable.

Like I said originally, I was talking about "managed value types" which are by definition non-blittable.

Reply:
> "managed value types" which are by definition non-blittable.

I've never heard of a "managed value type". Value types can have managed memory fields, but a value type can be on the stack (managed by the stack lifetime) or within an array (managed by the GC). Can you share where this term is used?

Reply:
> managed value type

Managed vs. unmanaged types is C# terminology: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/unmanaged-types

Reply:
> > managed value type
> 
> Managed vs. unmanaged types is C# terminology: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/unmanaged-types

Ah. I did not realize they can be applied to an entire value type. That was what I am missing. Thanks.

Reply:
> LayoutKind.Sequential vs. LayoutKind.Auto

C# marks structs as `LayoutKind.Sequential`. The `LayoutKind.Sequential` applies to the unmanaged view of the struct. It applies to the managed view of the struct only for blittable types where the managed and unmanaged view of struct are the same.

`EmbeddedValueType<string>` and `EmbeddedValueType<object>` types used by this test are not blittable, so the runtime is free to rearrange the fields in the managed view of the struct any way it wants. The current CoreCLR type loader policy is to place all object references first to improve locality for GC heap walking.

The comment "// Adding a simple field to force non-zero object offsets" does not match what is actually going to happen. I agree that it should be reworked.

Reply:
Tried to reach a compromise between all suggestions and added an explanatory comment.